### ISSUE TITLE
Fl_GIF_Image decoder bug

### ISSUE THREAD

**Issue by wcout** [fltk/fltk#274]
Stumbled upon a crash when trying load a GIF that was optimized by a tool.
The problem is with the LZW decompression, when the code table overflows.
A minor patch solves the issue:

````diff
diff --git a/src/Fl_GIF_Image.cxx b/src/Fl_GIF_Image.cxx
index d747ac480..8d58839df 100644
--- a/src/Fl_GIF_Image.cxx
+++ b/src/Fl_GIF_Image.cxx
@@ -363,15 +363,16 @@ void Fl_GIF_Image::load_gif_(Fl_Image_Reader &rdr)
     } while (tp > OutCode);
 
     if (OldCode != ClearCode) {
-      Prefix[FreeCode] = (short)OldCode;
-      Suffix[FreeCode] = FinChar;
+      if (FreeCode < 4096) {
+        Prefix[FreeCode] = (short)OldCode;
+        Suffix[FreeCode] = FinChar;
+      }
       FreeCode++;
       if (FreeCode > ReadMask) {
         if (CodeSize < 12) {
           CodeSize++;
           ReadMask = (1 << CodeSize) - 1;
         }
-        else FreeCode--;
       }
     }
     OldCode = CurCode;

````
The GIF doc says the application should continue to read 12 Bit Codes, but not store them. As it is now, the last code in the table gets overwritten, because FreeCode is decremented.

Here is my test gif:
[gif_lzw_bug.zip](https://github.com/fltk/fltk/files/7186735/gif_lzw_bug.zip)

**Comment by Albrecht-S** [comment]
Thanks for the report and the patch.

- Bug confirmed: test/pixmap_browser crashes when opening the provided GIF image whereas several other tools can open it
- Fix confirmed: the provided patch fixes this issue.

Before I commit this change I'd like to know more about the bug and the fix. Since you (@wcout) found the fix I'm sure you know much better what's going on than me (right now). Before I dig deeper into the code (this decoding algorithm looks pretty complicated), can you explain in a few more words why your fix is the correct one and - maybe even more important - how we can verify that it doesn't break other cases (w/o just testing lots of images)?

> The GIF doc says the application should continue to read 12 Bit Codes ...

Can you give a pointer to the docs, i.e. where do the docs say this? Is this an issue of the "Deferred Clear Code" mentioned in the beginning of [these GIF docs](https://www.w3.org/Graphics/GIF/spec-gif89a.txt)?

Thanks in advance!

**Comment by wcout** [comment]
The crash happens in line
https://github.com/fltk/fltk/blob/a774e120bc11dedc52c2e80185ccbd2f1acda59b/src/Fl_GIF_Image.cxx#L347
because tp advances unchecked beyond the buffer end.
Putting in a check like while (i >= ColorMapSize && tp < OutCode + 4096) fixes the crash, but the resulting image has artefacts.

Yes, I'm referring to this section in the GIF specification, the decoding part:

    As we can see from the above, when the decoder's table is full, it must
    not change the table until a clear code is received.  The Code Size is that
    of the maximum Code Size.  Processing other than this is done normally.

The current FLTK decoding changes the last table entry on this occasion, leading to the artefacts (and crash because the offsets go beyond 4096).

As I see it, this is an encoder issue also, because to me it makes no sense to store more "words in the dictionary" than the decoder can handle. But obviously other viewers can handle this situation correctly, so it seems an allowed thing.

**Comment by wcout** [comment]
> Before I dig deeper into the code (this decoding algorithm looks pretty complicated), can you explain in a few more words why your fix is the correct one

I can understand you are concerned about side effects of this change - this decrementing of FreeCount is 20 year old code (I've followed it back to a commit for FLTK 1.1!).
Nevertheless I'm quite certain that it is plainly wrong, as explained already.
I have now looked into several implementations that claim to be gif2ras descendants (like FL_GIF_Image). None of them has that decrement of FreeCount, although the rest of the code is still similar.
I think it maybe intended to secure the code from buffer overflows and did not realize it could change the dictionary table.

IMHO the code with this fix is completely in the intention of the specification: Act like the table were as big as FreeCode, but do not store new values in the table.

> how we can verify that it doesn't break other cases (w/o just testing lots of images)?

I fear apart from accurate code review from several people there is no other way than testing various images, like here:
https://github.com/robert-ancell/pygif/tree/master/test-suite

**Comment by Albrecht-S** [comment]
Thanks again for your comments. I agree that decrementing `FreeCode` is wrong.

Meanwhile I'm pretty sure that your fix is correct because I tried to understand the code and learned more about LZW (de)compression. There's one tiny bit I'd change though. It doesn't really matter in the end but it looks wrong in the resultant code after your patch: the increment of FreeCode should be inside the conditional code as follows:
```
    if (OldCode != ClearCode) {
      if (FreeCode < 4096) {
        Prefix[FreeCode] = (short)OldCode;
        Suffix[FreeCode] = FinChar;
        FreeCode++;
      }
```
Reasoning: it doesn't make sense to increment `FreeCode` if we don't store another value in the arrays. It wouldn't do any harm though because `FreeCode` would be reset when the next `ClearCode` is received.

> I fear apart from accurate code review from several people there is no other way than testing various images, like here: https://github.com/robert-ancell/pygif/tree/master/test-suite

Thanks for this link. Meanwhile I tested several hundred GIF images with success (not yet complete) and I'll download and test the images of this test suite as well.

Today I found another image that created an infinite loop when decoding with FLTK's decoder. Other image viewers issued an error message. It turned out that the image is incomplete (truncated) but we don't notice this fact; the effect is likely the same as with your test image posted at issue #271. **_We definitely need to fix both of these issues_** and I'm working on it.

> As I see it, this is an encoder issue also, because to me it makes no sense to store more "words in the dictionary" than the decoder can handle...

The effect is (as I understand it) that the encoder stores such "words" in the data stream as is, i.e. w/o compression, and the decoder reads them as-is. AFAICT this can make sense in some cases if the dictionary is filled with "good" codes and most of the new "words" can be compressed effectively rather than clearing the entire dictionary and building it again from fresh data. The encoder is obviously free to do that.

**Comment by Albrecht-S** [comment]
> I'll download and test the images of [this test suite](https://github.com/robert-ancell/pygif/tree/master/test-suite) as well.

Done. My current test version (which fixes the EOF/read error issue and outputs some test/debug data) can open all images in the `test-suite` folder whereas the old version would crash or loop when opening some of the images. I call this progress. ;-)
However, it's not yet production ready.

**Comment by wcout** [comment]
> You posted the "optimized" image, but can you also post the "original" (non-optimized) version for my analysis? Do you still have that original image?

Sort of. This was the 3rd frame of an optimized animated GIF. Here now is the same frame of the unoptimized animated GIF:
[gif_lzw_bug_original.zip](https://github.com/fltk/fltk/files/7200255/gif_lzw_bug_original.zip)

> Done. My current test version (which fixes the EOF/read error issue and outputs some test/debug data) can open all images in the test-suite folder whereas the old version would crash or loop when opening some of the images. I call this progress. ;-)

Great! Looking forward to see it.

**Comment by Albrecht-S** [comment]
Thanks for the original image.

FWIW: Here's the comparison of the output of my test code of both images:
```
[  128] Start decoding image 'gif_lzw_bug_original.gif'
[  130] Found ClearCode ( 32), FreeCode =   34,           fails = 0
[ 5769] Found ClearCode ( 32), FreeCode = 4096,           fails = 1
[11408] Found ClearCode ( 32), FreeCode = 4096,           fails = 2
[17047] Found ClearCode ( 32), FreeCode = 4096,           fails = 3
[22685] Found ClearCode ( 32), FreeCode = 4096,           fails = 4
[28324] Found ClearCode ( 32), FreeCode = 4096,           fails = 5
[31531] Found EOFCode   ( 33), FreeCode = 2481
[31531] EOF:    tpmax =  552,  Freecode = 4096 (max),     fails = 5

[  128] Start decoding image 'gif_lzw_bug.gif'   (the "optimized" image)
[  130] Found ClearCode ( 32), FreeCode =   34,           fails = 0
[ 5765] Found ClearCode ( 32), FreeCode = 4093,           fails = 0
[11399] Found ClearCode ( 32), FreeCode = 4093,           fails = 0
[17033] Found ClearCode ( 32), FreeCode = 4093,           fails = 0
[22667] Found ClearCode ( 32), FreeCode = 4093,           fails = 0
[29510] Found ClearCode ( 32), FreeCode = 4096,           fails = 801
[31098] Found EOFCode   ( 33), FreeCode = 1348
[31098] EOF:    tpmax =  518,  Freecode = 4096 (max),     fails = 801
```

- The first column `[...]` shows the corresponding file offsets and final sizes which differ significantly.
- `tpmax` is the maximal length of compressed items (552 vs 518).
- `FreeCode` shows the contents of the buffer at each given point.
- `fails` is a counter of items that couldn't be added to the buffer because it was full (`FreeCode == 4096`, i.e. "uncompressed" data).

The compressor buffer of the original image was obviously always filled up and cleared as soon as the first "failed" code word was found (`fails` increases by one). In the optimized image we can see that up to 801 "failed" entries are stored "uncompressed" before the buffer is cleared at file offset 29510.

**Comment by Albrecht-S** [comment]
This issue has been fixed in commit 1d847fe.

I consider this issue fixed. Test reports, feedback and suggestions for further improvement would be appreciated. TIA.
